src/main/java/org/helpberkeley/memberdata/HBParser.java

```diff
@@ -606,9 +606,19 @@ static void parseDetails(long postNumber, final String rawPost,
     }
 
     public static String shortURLDiscoursePost(final String line) {
-        int index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+        int index;
+        int prefixLength;
+
+        if (line.contains(Constants.UPLOAD_URI_PREFIX)) {
+            index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+            prefixLength= Constants.UPLOAD_URI_PREFIX.length();
+        }
+        else {
+            index = line.indexOf(Constants.WEB_CSV_PREFIX);
+            prefixLength= Constants.WEB_CSV_PREFIX.length();
+        }
         assert index != -1 : line;
```

Comment:
The original developer was being lazy here and making assumptions that the calling code would always be well behaved.  Could you please update this method,  shortURLUploadResponse(), and downloadFileName() replacing the assertion with an explicit check and throw MemberDataException if we haven't found the pattern we are looking for? Use one or more static final error messages and MessageFormat.format here and in the accompanying tests.  The tests should low level String based and live in a new test class called HBParserTest. Thank you.

Reply:
> Could you please update this method, shortURLUploadResponse(), and downloadFileName() replacing the assertion with an explicit check and throw MemberDataException if we haven't found the pattern we are looking for?

Hello David! While doing this, it occurred to me that it may be cleaner to add an assertion function in HBParser to check the prefix, and call that in all three of these functions. That way, if we ever wanted to change our supported file prefixes, we could just edit that function, instead of everywhere we need to check file prefixes. Additionally, of course, if we ever added more functions in HBParser that needed this functionality, we could just add a call to that function. Let me know what you think.

Reply:
Hello Kaelan.  Some stuff about assertions first. 

In stable running production code, assertions are typically disabled, for performance reasons.  They serve almost as documentation for assumptions being made by the code.  In the test environment, production code assertions are enabled. These are distinct from the test assert mechanisms in the test code.  Production code assertions are also sometimes enabled when trying to track down a production bug.

Assertions should not be used for dealing with user provided input, which can never be trusted to be well behaved. User provided input must always be explicitly checked and errors handled/reported appropriately.  Assertions are not a good mechanism for this because when triggered then cause an immediate error exit from the program.  In our case, running a server, that would mean the server crashing and not reporting anything back to the user.

All that is a long winded way of saying that rather than adding an assertion function to HBParser, let's talk about it as an auditing method, which would thrown an exception (MemberDataException) for problems.

Need to tend to something here at the moment.  Will get back to answering the gist of your suggestion in about 1/2 hour or so.

Reply:
Having a single place that encapsulates looking for an uploaded file URL is a good idea.  The UploadFile class is the container for this data, but parsing and creating it is spread out a bit due to organic growth within the code without cleanup pass in this area.  I think that we could hid this logic in a couple of static utility methods within UploadFile.  Something like:
```
   public static boolean containsUploadedFileURL(String data);

    // throws MemberDataException if it can't find a valid URL
   public static UploadFile createUploadFile(String data);

Take a look through the usages of UploadFile throughout the code and let me know what you think.